framework/Session_menu.cpp

```diff
@@ -277,6 +277,15 @@ idSessionLocal::SetPbMenuGuiVars
 void idSessionLocal::SetPbMenuGuiVars( void ) {
 }
 
+void idSessionLocal::ToggleMenu() {
+	if ( guiActive != guiMainMenu ) {
+		console->Close();
+		StartMenu();
+	} else if ( mapSpawned ) {
+		ExitMenu();
+	}
```

Comment:
This is what I originally did, hardcoding the gamepad button to K_ESCAPE, but it's pretty terrible and I want to avoid it.

To start with, imho hardcoding the ESCAPE key is bad and a hack in the first place (completely circumventing the otherwise customizable input system), so I don't want to further cement this as I'd rather see it refactored some time in the future. Secondly, emitting a K_ESCAPE event for a gamepad button is another hack on top of that which creates problems for bindability of the gamepad button in question. In fact, my original solution just hardcoded the START button to K_ESCAPE, which also prevented it from binding long press or modifier actions etc.. So either way, I need a command that the button can be bound to, and imho having a command `emulateEscape` is just weird.

If you are concerned about code duplication, I'd rather change the existing handling of the ESC key in the GUI context to the new command than the other way around.

Reply:
Yes, my suggestion is to turn it into "escape" key after all the binds. Instead of binding something to "togglemenu", bind it to "escapekey" or "seEscape", which generates escape key press.

I'm not sure there is anything to refactor. `sysEvent_t` sounds like a system event, not raw key press. The fact that it reuses key codes for Escape, functional keys, and maybe something else is probably 1) for convenience, 2) because there was no other input device supporting that many buttons anyway.

Reply:
I disagree. `sysEvent_t` is just a vehicle to transport certain events around the engine, and it has a specific `type` field that can be `SE_KEY`, `SE_MOUSE` or other things. The check is specifically for an `SE_KEY` event of type `K_ESCAPE`, and only keyboard input devices would ever generate such an event. This is not convenience, this is a hardcoded key mapping.
Worse, it's used in several places in UI code for different things that don't really make sense to have on a gamepad bind. While thankfully not really harmful, it just muddies the water.
Finally, the method to generate any such `sysEvent_t` event was actually internal to the `Sys_` platform functions and not publicly exposed before this PR. So injecting this kind of event externally is definitely a hack.

But whatever. I've changed it as you requested. After all, why would we want to clean up UI interactions?

Reply:
It was possible to generate sysEvent_t before, like this:
```
void Automation::PushInputEvents(int key, int value) {
	sysEvent_t ev;
	memset(&ev, 0, sizeof(ev));
	ev.evType = SE_KEY;
	ev.evValue = key;
	ev.evValue2 = value;
	eventLoop->PushEvent(&ev);
}
```
I think it is not the only such key. I'm pretty sure arrows, delete, and backspace in GUI code are also hardcoded to K_XXX constants.